UI node outlines

ickshonpe opened the pull request:
# Objective

Add support for drawing outlines outside the borders of UI nodes.

## Solution
Add a new `Outline` component with `width`, `offset` and `color` fields. 
Added `outline_width` and `outline_offset` fields to `Node`. This is set after layout recomputation by the `resolve_outlines_system`.

Properties of outlines:
* Unlike borders, outlines have to be the same width on each edge.
* Outlines do not occupy any space in the layout.
* The `Outline` component won't be added to any of the UI node bundles, it needs to be inserted separately. 
* Outlines are drawn outside the node's border, so they are clipped using the clipping rect of their entity's parent UI node (if it exists).
*  `Val::Percent` outline widths are resolved based on the width of the outlined UI node.
* The offset of the `Outline` adds space between an outline and the edge of its node.

I was leaning towards adding an `outline` field to `Style` but a separate component seems more efficient for queries and change detection. The `Outline` component isn't added to bundles for the same reason.

---
## Examples

* This image is from the `borders` example from the Bevy UI examples but modified to include outlines. The UI nodes are the dark red rectangles, the bright red rectangles are borders and the white lines offset from each node are the outlines. The yellow rectangles are separate nodes contained with the dark red nodes:
 
   <img width="406" alt="outlines" src="https://github.com/bevyengine/bevy/assets/27962798/4e6f315a-019f-42a4-94ee-cca8e684d64a">

* This is from the same example but using a branch that implements border-radius. Here the the outlines are in orange and there is no offset applied. I broke the borders implementation somehow during the merge, which is why some of the  borders from the first screenshot are missing :sweat_smile:. The outlines work nicely though (as long as you can forgive the lack of anti-aliasing):

   ![image](https://github.com/bevyengine/bevy/assets/27962798/d15560b6-6cd6-42e5-907b-56ccf2ad5e02)

---
## Notes

As I explained above, I don't think the `Outline` component should be added to UI node bundles. We can have helper functions though, perhaps something as simple as:

```rust
impl NodeBundle {
    pub fn with_outline(self, outline: Outline) -> (Self, Outline) {
        (self, outline)
    }
}
```

I didn't include anything like this as I wanted to keep the PR's scope as narrow as possible. Maybe `with_outline` should be in a trait that we implement for each UI node bundle. 

---

## Changelog
Added support for outlines to Bevy UI.
* The `Outline` component adds an outline to a UI node.
* The `outline_width` field added to `Node` holds the resolved width of the outline, which is set by the `resolve_outlines_system` after layout recomputation.
* Outlines are drawn by the system `extract_uinode_outlines`.

nicoburns reviewed (approved):
Seems useful as-is, but I wonder how hard it would be to add https://developer.mozilla.org/en-US/docs/Web/CSS/outline-offset support to this? Could be useful for people who want to draw outlines inside the node boundaries rather than outside.

ickshonpe commented:
> Seems useful as-is, but I wonder how hard it would be to add https://developer.mozilla.org/en-US/docs/Web/CSS/outline-offset support to this? Could be useful for people who want to draw outlines inside the node boundaries rather than outside.

Really trivial, I'll add it now

ickshonpe commented:
Also got outlines working with border-radius as well:

<img width="459" alt="curved_outlines" src="https://github.com/bevyengine/bevy/assets/27962798/4a746458-bdc5-4a64-97e5-5bbbd4215547">

JoJoJet commented on crates/bevy_ui/src/layout/mod.rs:
```diff
@@ -383,6 +383,34 @@ pub fn ui_layout_system(
+        / ui_scale.0 as f32;
+
+    for (outline, mut node) in outlines_query.iter_mut() {
+        let node = node.bypass_change_detection();
```
How come you're calling `bypass_change_detection` here?

ickshonpe replied:
The text system queries for changes to `Node` so it can recompute the text layout on changes to the UI node's size.

JoJoJet replied:
So, do we not care about detecting changes to the `outline_width` and `outline_offset` fields?

ickshonpe replied:
Not at the moment. Maybe we could have a separate component for every property but it doesn't seem very tractable.

JoJoJet replied:
Fair enough. I'd put a comment here describing that we don't care about change detection for these fields because they're private and only used as a cache.

wilk10 commented:
Love this feature! 

I have one question about the implementation: one of the main use-cases I can see would be to show the outline on the currently hovered element in a list or grid. In that case, as a user I would either be removing and adding the `Outline` component very frequently, or I guess I can set the `width`to 0 or the value I want.

IMO it would be more explicit and discoverable to have an `is_visible` flag as field on `Outline` and add `Outline` to all relevant node bundles, so that then I can just toggle it. What do you think?

nicopap commented:
I think setting width to `0` is clear enough as is. Maybe adding a doc line saying "Consider setting the size to zero if you are dynamically changing the outline width" could fill the gap. Also the current node bundles are already way too large, and results in a lot of extra machine codes. It's reasonable to want to keep them "tight".

ickshonpe commented:
> Love this feature!
> 
> I have one question about the implementation: one of the main use-cases I can see would be to show the outline on the currently hovered element in a list or grid. In that case, as a user I would either be removing and adding the `Outline` component very frequently, or I guess I can set the `width`to 0 or the value I want.
> 
> IMO it would be more explicit and discoverable to have an `is_visible` flag as field on `Outline` and add `Outline` to all relevant node bundles, so that then I can just toggle it. What do you think?

Setting the outline's color to `Color::NONE` makes the most sense to me for toggling outlines on and off. An `is_visible` method feels like it would be too easily confused with Bevy's existing visibility systems. 

Most UI nodes won't use outlines, so it wouldn't be efficient to add it to every bundle. If you do need to create a lot of outlined bundles you can always use a builder method or a type alias for `(NodeBundle, Outline)`.

wilk10 commented:
> > Love this feature!
> > I have one question about the implementation: one of the main use-cases I can see would be to show the outline on the currently hovered element in a list or grid. In that case, as a user I would either be removing and adding the `Outline` component very frequently, or I guess I can set the `width`to 0 or the value I want.
> > IMO it would be more explicit and discoverable to have an `is_visible` flag as field on `Outline` and add `Outline` to all relevant node bundles, so that then I can just toggle it. What do you think?
> 
> Setting the outline's color to `Color::NONE` makes the most sense to me for toggling outlines on and off. An `is_visible` method feels like it would be too easily confused with Bevy's existing visibility systems.
> 
> Most UI nodes won't use outlines, so it wouldn't be efficient to add it to every bundle. If you do need to create a lot of outlined bundles you can always use a builder method or a type alias for `(NodeBundle, Outline)`.

Thanks for the reply. I understand and setting the color to `Color::NONE` should work and be clear enough. Does it make sense to add methods on `Outline`to `show` / `hide` (or `draw` or other alternative names) to control `color`, in order to point the user in the right direction? Maybe at least a comment on the `color` field?

My main concern regards table moves when the `Outline` component is added / removed frequently (eg: an inventory system implemented as a grid, where only the hovered element needs an outline). Setting the color to `Color::NONE` is clearly the better solution than inserting / removing the component, IMO.

ickshonpe commented:
> > > Love this feature!
> > > I have one question about the implementation: one of the main use-cases I can see would be to show the outline on the currently hovered element in a list or grid. In that case, as a user I would either be removing and adding the `Outline` component very frequently, or I guess I can set the `width`to 0 or the value I want.
> > > IMO it would be more explicit and discoverable to have an `is_visible` flag as field on `Outline` and add `Outline` to all relevant node bundles, so that then I can just toggle it. What do you think?
> > 
> > 
> > Setting the outline's color to `Color::NONE` makes the most sense to me for toggling outlines on and off. An `is_visible` method feels like it would be too easily confused with Bevy's existing visibility systems.
> > Most UI nodes won't use outlines, so it wouldn't be efficient to add it to every bundle. If you do need to create a lot of outlined bundles you can always use a builder method or a type alias for `(NodeBundle, Outline)`.
> 
> Thanks for the reply. I understand and setting the color to `Color::NONE` should work and be clear enough. Does it make sense to add methods on `Outline`to `show` / `hide` (or `draw` or other alternative names) to control `color`, in order to point the user in the right direction? Maybe at least a comment on the `color` field?
> 
> My main concern regards table moves when the `Outline` component is added / removed frequently (eg: an inventory system implemented as a grid, where only the hovered element needs an outline). Setting the color to `Color::NONE` is clearly the better solution than inserting / removing the component, IMO.

I wouldn't consider the API here final by any means. 

I understand your concerns about table moves. As well as the performance issues, at the moment in Bevy UI they can affect the order of UI elements atm in some circumstances (this should be fixed soon though, we've fixed some of the problems already, and I've been working on changes that should fix the remaining issues). 

Show/Hide helper methods have similar problems to the `is_visible` method though, it feels like they could be confused with the `Style` property settings.  I added what you suggested  the `color` field's description and some examples to the struct's doc comments that should help I think, also maybe we could add an outlined buttons example in a separate PR.

alice-i-cecile commented:
@ickshonpe once merge conflicts are resolved I'll merge this.

ickshonpe commented:
this is ready to merge
